lib/maths/analytics/CDataFrameCategoryEncoder.cc

```diff
@@ -239,6 +231,74 @@ class CMinRedundancyMaxRelevancyGreedySearch {
     double m_RedundancyWeight;
     TFeatureRelevanceMinusRedundancyList m_Features;
 };
+
+template<typename T>
+void getEncodingAttribute(const json::object& obj, const std::string& tag, T& value) {
+    if (obj.contains(tag) == false) {
+        throw std::runtime_error("Expected attribute [\"" + tag +
+                                 "\"] not found in : " + json::serialize(obj));
+    }
+    if (obj.at(tag).is_string() == false) {
+        throw std::runtime_error("Value at [\"" + tag +
+                                 "\"] is not a string: : " + json::serialize(obj));
+    }
+    // We use json::value_to to convert from json::string type to std::string
+    if (core::CStringUtils::stringToType(json::value_to<std::string>(obj.at(tag)),
+                                         value) == false) {
+        throw std::runtime_error("Unable to convert value at [\"" + tag +
+                                 "\"] to requested type: " + json::serialize(obj));
+    }
+}
+
+const json::object& extractColumnIndexAndMic(const json::value& encodingObj,
+                                             const std::string& tag,
+                                             std::size_t& colIdx,
+                                             double& mic) {
+    std::ostringstream err;
+    if (encodingObj.as_object().at(tag).is_object() == false) {
+        err << encodingObj.as_object().at(tag);
+        throw std::runtime_error("JSON value is not an object: " + err.str());
+    }
+    const json::object& obj = encodingObj.as_object().at(tag).as_object();
+    LOG_TRACE(<< tag << " obj: " << obj);
+
+    getEncodingAttribute(obj, ENCODING_INPUT_COLUMN_INDEX_TAG, colIdx);
+    getEncodingAttribute(obj, ENCODING_MIC_TAG, mic);
+
+    return std::cref(obj);
+}
+
+void extractMappedEncodingAndFallback(const json::object& obj, TDoubleVec& map, double& fallback) {
+    std::ostringstream err;
+    const json::value& encodingMap = obj.at(MAPPED_ENCODING_MAP_TAG);
+    if (encodingMap.is_string() == false) {
+        err << encodingMap;
+        throw std::runtime_error("JSON value is not a string: " + err.str());
+    }
+    core::CStringUtils::TStrVec tokens;
+    std::string rmdr;
+    core::CStringUtils::tokenise(":", encodingMap.as_string().c_str(), tokens, rmdr);
```

Comment:
Longer term (not necessary for this PR) we should make `CStringUtils::tokenise` take a `std::string_view` as input. That would avoid the inefficiency here of creating a second `std::string` in addition to Boost.JSON's "stringlike" object.

Reply:
Noted